python/ray/data/_internal/logical/interfaces.py

```diff
@@ -0,0 +1,123 @@
+from typing import List
+
+from ray.data._internal.execution.interfaces import Operator, PhysicalOperator
+from ray.data._internal.execution.legacy_compat import (
+    _blocks_to_input_buffer,
+    _stage_to_operator,
+)
+from ray.data._internal.plan import Stage
+
+
+class LogicalOperator(Operator):
+    """Abstract class for logical operators.
+
+    A logical operator describes transformation, and later is converted into
+    physical operator.
+    """
+
+    def __init__(self, name: str, input_dependencies: List["LogicalOperator"]):
+        super().__init__(name, input_dependencies)
+        for x in input_dependencies:
+            assert isinstance(x, LogicalOperator), x
+
+    def set_input_dependencies(self, input_dependencies: List["LogicalOperator"]):
+        self._input_dependencies = input_dependencies
+
+    def set_legacy_stage(self, stage: Stage):
```

Comment:
The legacy stage stuff is a bit awkward. Could we break the link so that at least the new operator copies all necessary fields without needing a stage reference?

Reply:
Yeah, was meant to do a quick prototype. Removed now.